### PR TITLE
ASoC: SOF: intel: don't read CTX_SAVE reply

### PR DESCRIPTION
**Pull request by bardliao**
The memory windows are powered off before sending IPC reply back to the
host, so the host should not try to read the mailbox for CTX_SAVE reply.

Signed-off-by: Bard liao <yung-chuan.liao@linux.intel.com>

Fix https://github.com/thesofproject/sof/issues/984

### CODE REVIEW

File: sound/soc/sof/intel/hda-ipc.c [hunk 1]
```diff
@@ -89,8 +89,19 @@ int hda_dsp_ipc_get_reply(struct snd_sof_dev *sdev,
 	int ret = 0;
 	u32 size;
 
-	/* get IPC reply from DSP in the mailbox */
-	sof_mailbox_read(sdev, sdev->host_box.offset, &reply, sizeof(reply));
+	if (msg->header == (SOF_IPC_GLB_PM_MSG | SOF_IPC_PM_CTX_SAVE)) {
```
Comment (plbossart) [inline sound/soc/sof/intel/hda-ipc.c]: I thought we discussed that the msg->header field would no longer be used, see e.g. https://github.com/thesofproject/linux/pull/586
May be that's a good counter-example that it should? Or can you use the body of the msg to find thhe command information

**Comment by ranj063** [comment]
@bardliao should we generalize this to maybe support other IPC messages that dont require reading the mailbox?

File: sound/soc/sof/intel/hda-ipc.c (hunk 1 above)
Comment (bardliao) [reply sound/soc/sof/intel/hda-ipc.c]: @plbossart Yes, I can use the body to find the command information as ```struct sof_ipc_cmd_hdr``` is always the first element of ```msg_data```.  However, I think check ```msg->header``` doesn't against #586.  We don't write ``msg->header`` to SHIM register in #586, but it doesn't touch ``msg->header``. In other word, we can still use ``msg->header`` after #586 is merged.

**Comment by bardliao** [comment]
> @bardliao should we generalize this to maybe support other IPC messages that dont require reading the mailbox?

@ranj063 Yes, it's a good idea. Probably to add a flag in ``struct snd_sof_ipc_msg``? But I am thinking is it worth to do it?  We will still need to check if the message is a non-reply one when we set the flag.

**Comment by plbossart** [comment]
On 2/14/19 8:33 PM, Bard Liao wrote:
>
> *@bardliao* commented on this pull request.
>
> ------------------------------------------------------------------------
>
> In sound/soc/sof/intel/hda-ipc.c 
> <https://github.com/thesofproject/linux/pull/651#discussion_r257087314>:
>
> > @@ -89,8 +89,19 @@ int hda_dsp_ipc_get_reply(struct snd_sof_dev *sdev,
>   	int ret = 0;
>   	u32 size;
>   
> -	/* get IPC reply from DSP in the mailbox */
> -	sof_mailbox_read(sdev, sdev->host_box.offset, &reply, sizeof(reply));
> +	if (msg->header == (SOF_IPC_GLB_PM_MSG | SOF_IPC_PM_CTX_SAVE)) {
>
> @plbossart <https://github.com/plbossart> Yes, I can use the body to 
> find the command information as |struct sof_ipc_cmd_hdr| is always the 
> first element of |msg_data|. However, I think check |msg->header| 
> doesn't against #586 
> <https://github.com/thesofproject/linux/pull/586>. We don't write 
> |msg->header| to SHIM register in #586 
> <https://github.com/thesofproject/linux/pull/586>, but it doesn't 
> touch |msg->header|. In other word, we can still use |msg->header| 
> after #586 <https://github.com/thesofproject/linux/pull/586> is merged.
>
it's really the other way around. If we no longer use the msg->header to 
pass any information to the firmware, then the plan was to remove this 
field since the information it contained was duplicated from the body of 
the message going over a mailbox.

That said, there may be a need for sideband signals for power management 
that don't go over the mailbox. I'd like the thread of the IPC design to 
make progress before we commit to using this field or remove it.

File: sound/soc/sof/intel/hda-ipc.c (hunk 1 above)
Comment (plbossart) [reply sound/soc/sof/intel/hda-ipc.c]: no the idea was that the msg->header would be removed if it serves no purpose.

Comment (bardliao) [reply sound/soc/sof/intel/hda-ipc.c]: @plbossart updated. use the body of the msg to check if it is SOF_IPC_PM_CTX_SAVE

**Comment by emilchudzik** [comment]
Playback and suspend works fine in loop with applied this patch. There is no error message with it.

kernel: cros-sof-v4.19 / [8492020d5967](https://github.com/thesofproject/linux/commit/8492020d5967)
FW: cml-002-drop-stable / [dcd825b](https://github.com/thesofproject/sof/commit/dcd825b923a680f63adde1b392638a14463ead63)
tplg for hda: sof-hda-generic.tplg

File: sound/soc/sof/intel/hda-ipc.c [hunk 2]
```diff
@@ -86,11 +86,25 @@ int hda_dsp_ipc_get_reply(struct snd_sof_dev *sdev,
...
 
-	/* get IPC reply from DSP in the mailbox */
-	sof_mailbox_read(sdev, sdev->host_box.offset, &reply, sizeof(reply));
+
+	hdr = (struct sof_ipc_cmd_hdr *)msg->msg_data;
+	if (hdr->cmd == (SOF_IPC_GLB_PM_MSG | SOF_IPC_PM_CTX_SAVE)) {
+		/*
+		 * memory windows are powered off before sending IPC reply,
+		 * so we can't read the mailbox for CTX_SAVE reply.
+		 */
```
Comment (plbossart) [inline sound/soc/sof/intel/hda-ipc.c:98]: Isn't the header in a memory window? How does this work if the memory is no longer powered?

It feels like we actually do need to preserve the use of the msg/msg_ext bits precisely for this case.

Comment (bardliao) [reply sound/soc/sof/intel/hda-ipc.c:98]: @plbossart msg->msg_data is dynamic allocated [msg->msg_data = devm_kzalloc(sdev->dev, PAGE_SIZE, GFP_KERNEL);](https://github.com/thesofproject/linux/blob/topic/sof-dev/sound/soc/sof/ipc.c#L816). I think it is not in a memory window.

**Comment by xun2z** [comment]
I test the newest patch on WHL RVP, here is the dmesg, no CTX_SAVE error message.

**Test Env:**
kernel: cros-sof-v4.19 8492020 (with patch pr651)
sof: cml-002-drop-stable dcd825b
tplg: sof-whl-rt5682.tplg

dmesg:
```
[   51.667062] sof-audio-pci 0000:00:1f.3: ipc tx: 0x40020000: GLB_PM_MSG: CTX_RESTORE
[   51.667198] sof-audio-pci 0000:00:1f.3: ipc tx succeeded: 0x40020000: GLB_PM_MSG: CTX_RESTORE
[   51.667205] HDMI HDA Codec ehdaudio0D2: hdac_hdmi_present_sense: disconnect for pin:port 6:0
[   51.667207] HDMI HDA Codec ehdaudio0D2: hdac_hdmi_present_sense: disconnect for pin:port 7:0
[   51.667288] serial 00:06: activated
[   51.831824] ata3: SATA link down (SStatus 4 SControl 300)
[   51.897655] atkbd serio0: Failed to deactivate keyboard on isa0060/serio0
[   51.964830] acpi LNXPOWER:08: Turning OFF
[   51.965354] acpi LNXPOWER:07: Turning OFF
[   51.967427] OOM killer enabled.
[   51.967428] Restarting tasks ... done.
[   52.071071] video LNXVIDEO:00: Restoring backlight state
[   52.071075] PM: suspend exit
[   52.164210] sof-audio-pci 0000:00:1f.3: ipc rx: 0x90020000: GLB_TRACE_MSG
[   52.164293] sof-audio-pci 0000:00:1f.3: ipc rx done: 0x90020000: GLB_TRACE_MSG
[   52.281670] atkbd serio0: Failed to enable keyboard on isa0060/serio0
[   52.664314] sof-audio-pci 0000:00:1f.3: ipc rx: 0x90020000: GLB_TRACE_MSG
[   52.664399] sof-audio-pci 0000:00:1f.3: ipc rx done: 0x90020000: GLB_TRACE_MSG
[   53.068840] sof-audio-pci 0000:00:1f.3: ipc tx: 0x40010000: GLB_PM_MSG: CTX_SAVE
[   53.069112] sof-audio-pci 0000:00:1f.3: ipc tx succeeded: 0x40010000: GLB_PM_MSG: CTX_SAVE
[   53.069157] sof-audio-pci 0000:00:1f.3: FW Poll Status: reg=0x1010f0f successful
[   53.069729] sof-audio-pci 0000:00:1f.3: FW Poll Status: reg=0xf0f successful
[   53.069742] sof-audio-pci 0000:00:1f.3: DSP core(s) enabled? 0 : core_mask f
[   53.069817] sof-audio-pci 0000:00:1f.3: Debug PCIR: 00000000 at  00000044
[   53.069826] sof-audio-pci 0000:00:1f.3: Debug PCIW: 00000010 at  00000044
[   53.587822] e1000e: enp0s31f6 NIC Link is Up 100 Mbps Full Duplex, Flow Control: None
[   53.587825] e1000e 0000:00:1f.6 enp0s31f6: 10/100 speed: disabling TSO
```
[dmesg.txt](https://github.com/thesofproject/linux/files/2892861/1.txt)

**Comment by plbossart** [comment]
> +	hdr = (struct sof_ipc_cmd_hdr *)msg->msg_data;
> +	if (hdr->cmd == (SOF_IPC_GLB_PM_MSG | SOF_IPC_PM_CTX_SAVE)) {
> +		/*
> +		 * memory windows are powered off before sending IPC reply,
> +		 * so we can't read the mailbox for CTX_SAVE reply.
> +		 */
>
> @plbossart <https://github.com/plbossart> msg->msg_data is dynamic 
> allocated msg->msg_data = devm_kzalloc(sdev->dev, PAGE_SIZE, 
> GFP_KERNEL); 
> <https://github.com/thesofproject/linux/blob/topic/sof-dev/sound/soc/sof/ipc.c#L816>. 
> I think it is not in a memory window.
>
That doesn't seem right. where is the data actually coming from? 
Everything we do is based on memory-windows so there's got to be a copy 
somewhere.
>

**Comment by bardliao** [comment]
> That doesn't seem right. where is the data actually coming from? Everything we do is based on memory-windows so there's got to be a copy somewhere.

The data is from [memcpy(msg->msg_data, msg_data, msg_bytes);](https://github.com/thesofproject/linux/blob/topic/sof-dev/sound/soc/sof/ipc.c#L268)
And the source is from https://github.com/thesofproject/linux/blob/topic/sof-dev/sound/soc/sof/pm.c#L198

**Comment by ranj063** [comment]
> +	hdr = (struct sof_ipc_cmd_hdr *)msg->msg_data; +	if (hdr->cmd == (SOF_IPC_GLB_PM_MSG | SOF_IPC_PM_CTX_SAVE)) { +	/* +	* memory windows are powered off before sending IPC reply, +	* so we can't read the mailbox for CTX_SAVE reply. +	*/ @plbossart <https://github.com/plbossart> msg->msg_data is dynamic allocated msg->msg_data = devm_kzalloc(sdev->dev, PAGE_SIZE, GFP_KERNEL); <https://github.com/thesofproject/linux/blob/topic/sof-dev/sound/soc/sof/ipc.c#L816>. I think it is not in a memory window.
> That doesn't seem right. where is the data actually coming from? Everything we do is based on memory-windows so there's got to be a copy somewhere.
> […](#)

@plbossart this is not coming from the memory window. The msg_data is pre-allocated kernel memory which is used for copying to the mailbox during ipc tx. The mailbox reply reading step actually happens after this check. So this check looks legitimate. 

But my concern is about the placement of this check. Should we possibly move this check to tx_wait_done() prior to https://github.com/thesofproject/linux/blob/ee19d24e26c7e8cb193bfb72cce27561ec281c92/sound/soc/sof/ipc.c#L221. @bardliao what do you think?

While I'm at it, I want to raise my original point again. I think this should be more generic by maybe adding a field/flag to the ipc msg structure which suggests that the reply should not be read for this msg.

**Comment by bardliao** [comment]
@plbossart CI issues are solved

**Pull request merged**